hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/multipart/S3MultipartUploadCompleteRequest.java

```diff
@@ -285,7 +285,7 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
           // But now as versioning is not supported, just following the commit
           // key approach. When versioning support comes, then we can uncomment
           // below code keyInfo.addNewVersion(locations);
-          omKeyInfo.updateLocationInfoList(partLocationInfos, true);
+          omKeyInfo.updateLocationInfoList(partLocationInfos, true, true);
```

Comment:
can you elaborate why we need to skip block id check for s3 MPU compelte?

Reply:
Thanks @xiaoyuyao for the comment.

Please refer [MPU test case](https://github.com/apache/ozone/blob/master/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java#L2362), here MPU key has been overridden and the existing OM behavior is that it allows updated block locations of an existing MPU key. Thats the reason for skipping the blockID check on S3MultipartUploadCompleteRequest.

Reply:
Blindly allow recommit via MPU can be problematic as we don't have version support. The overwritten data will be leaked and not garbage collected properly. But this can be solved with a separate JIRA. cc: @bharatviswa504

Reply:
I had the same question / doubts and tried to understand the code (and it LGTM):

Based on my understanding in case of MPU the keys are created during the `S3MultipartUploadCommit` call. This call creates an entry in the `MultiPartInfoTable` which includes the key location information. 

During the complete phase these (already validated !) key location information are moved from the `MultipartInfo` entries to a final key entry.

The `MultipartInfo` entries seem to be properly validated in `S3MultipartUploadCommitPartRequest` L:160:

```
 omKeyInfo.updateLocationInfoList(keyArgs.getKeyLocationsList().stream()
          .map(OmKeyLocationInfo::getFromProtobuf)
          .collect(Collectors.toList()), true); 
```

This is the default version (with validation):

```
  */
  public void updateLocationInfoList(List<OmKeyLocationInfo> locationInfoList,
      boolean isMpu) {
    updateLocationInfoList(locationInfoList, isMpu, false);
  }
```

--> Based on my understanding everything is perfect with this patch. As far as the good part infos are used, the key information is already validated. (and the part information is guarded by the randomness of the uploadId:`UUID.randomUUID().toString() + "-" + UniqueId.next()` which is known only by the client)

Reply:
>The overwritten data will be leaked and not garbage collected properly.
It is already taken care in S3MultipartUploadCommitRequest and S3MultipartUploadCommitPartResponse.
The over-ride part is moved to delete Table.

Reply:
Thanks @bharatviswa504  and @elek , that makes sense to me.

Reply:
Thanks @xiaoyuyao , @bharatviswa504 , @elek for the useful comments.